mbedtls/src/bignum/mod.rs

```diff
@@ -237,6 +237,148 @@ impl Mpi {
         Ok(r)
     }
 
+    /// Returns an integer a such that (a*a) === self (mod p)
+    ///
+    /// The modulus must be prime; computing a square root modulo
+    /// a composite number is equivalent to factoring the composite.
+    pub fn mod_sqrt(&self, p: &Mpi) -> ::Result<Mpi> {
+        let zero = Mpi::new(0)?;
+
+        if self < &zero || self >= p {
+            return Err(::Error::MpiBadInputData);
+        }
+        if self == &zero {
+            return Ok(zero);
+        }
+
+        // This ignores p=2 (for which this algorithm is valid), as not cryptographically interesting.
+        if p.get_bit(0) == false || p <= &zero {
+            return Err(::Error::MpiBadInputData);
+        }
+
+        if self.jacobi(p)? != 1 {
+            // a is not a quadratic residue mod p
+            return Err(::Error::MpiBadInputData);
+        }
+
+        if (p % 4)?.as_u32()? == 3 {
+            // simple case for which we can avoid Shanks-Tonelli
+            let exp = ((p + 1)? >> 2)?;
+            return self.mod_exp(&exp, p);
+        }
+
+        /*
+        The case p % 4 == 1 requires Shanks-Tonelli algorithm
+
+        See http://www.math.vt.edu/people/brown/class_homepages/shanks_tonelli.pdf
+        */
+        let p_minus_1 = (p - 1)?;
+        let e = p_minus_1.trailing_zeros();
+        let s = (p_minus_1 >> e)?;
+
+        /*
+        Find a random quadratic nonresidue mod p.
+        */
+        let mut n = Mpi::new(2)?;
+        while n.jacobi(p)? != -1 {
+            n += 1;
+        }
+
+        let mut x = self.mod_exp(&(&(&s + 1)? >> 1)?, p)?;
+        let mut b = self.mod_exp(&s, p)?;
+        let mut g = n.mod_exp(&s, p)?;
+        let mut r = e;
+
+        let one = Mpi::new(1)?;
+        let two = Mpi::new(2)?;
+
+        loop {
+            if b == one {
+                return Ok(x); // and p-x
+            }
+
+            // Find by repeated squaring first m such that b^(2*m) == 1 mod p
+            let mut m = 0;
+
+            let mut bo = b.clone();
+            while bo != one {
+                bo = bo.mod_exp(&two, p)?;
+                m += 1;
+                if m >= r {
+                    return Err(::Error::MpiBadInputData);
+                }
+            }
+
+            // g^(2^(r-m-1))
+            let exp = (&one << (r - m - 1))?;
+            let t = g.mod_exp(&exp, p)?;
+            // g^(2^(r-m))
+            let t2 = t.mod_exp(&two, p)?;
+
+            x = (&x * &t)?.modulo(p)?;
+            b = (&b * &t2)?.modulo(p)?;
+            g = t2;
+            r = m;
+        }
+    }
+
+    fn trailing_zeros(&self) -> usize {
```

Comment:
`pub fn`?

Reply:
Might as well. We didn't need it to be public in the internal crate but this operation comes up a lot in number theoretic algorithms (as can be seen here by it being used in both Jacobi and Shanks-Tonelli functions) and a user of the crate might have need for it.